hcl2template/types.packer_config.go

```diff
@@ -42,6 +42,118 @@ func (cfg *PackerConfig) EvalContext() *hcl.EvalContext {
 	return ectx
 }
 
+// decodeInputVariables looks in the found blocks for 'variables' and
+// 'variable' blocks. It should be called firsthand so that other blocks can
+// use the variables.
+func (c *PackerConfig) decodeInputVariables(f *hcl.File) hcl.Diagnostics {
+	var diags hcl.Diagnostics
+
+	content, moreDiags := f.Body.Content(configSchema)
+	diags = append(diags, moreDiags...)
+
+	for _, block := range content.Blocks {
+		switch block.Type {
+		case variableLabel:
+			moreDiags := c.InputVariables.decodeVariableBlock(block, nil)
+			diags = append(diags, moreDiags...)
+		case variablesLabel:
+			attrs, moreDiags := block.Body.JustAttributes()
+			diags = append(diags, moreDiags...)
+			for key, attr := range attrs {
+				moreDiags = c.InputVariables.decodeVariable(key, attr, nil)
+				diags = append(diags, moreDiags...)
+			}
+		}
+	}
+	return diags
+}
+
+// parseLocalVariables looks in the found blocks for 'locals' blocks. It
+// should be called after parsing input variables so that they can be
+// referenced.
+func (c *PackerConfig) parseLocalVariables(f *hcl.File) ([]*Local, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+
+	content, moreDiags := f.Body.Content(configSchema)
+	diags = append(diags, moreDiags...)
+	var allLocals []*Local
+
+	for _, block := range content.Blocks {
+		switch block.Type {
+		case localsLabel:
+			attrs, moreDiags := block.Body.JustAttributes()
+			diags = append(diags, moreDiags...)
+			locals := make([]*Local, 0, len(attrs))
+			for name, attr := range attrs {
+				if _, found := c.LocalVariables[name]; found {
+					diags = append(diags, &hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  "Duplicate variable",
+						Detail:   "Duplicate " + name + " variable definition found.",
+						Subject:  attr.NameRange.Ptr(),
+						Context:  block.DefRange.Ptr(),
+					})
+					return nil, diags
+				}
+				locals = append(locals, &Local{
+					Name: name,
+					Expr: attr.Expr,
+				})
+			}
+
+			if allLocals == nil {
+				allLocals = locals
+			} else {
+				allLocals = append(allLocals, locals...)
+			}
+		}
+	}
+
+	return allLocals, diags
+}
+
+func (c *PackerConfig) evaluateLocalVariables(locals []*Local) hcl.Diagnostics {
+	var diags hcl.Diagnostics
+
+	if len(locals) > 0 && c.LocalVariables == nil {
+		c.LocalVariables = Variables{}
+	}
+
+	for i, local := range locals {
+		for _, transversal := range local.Expr.Variables() {
+			if transversal.RootName() == "local" {
+				// Evaluate later the locals that use another local variable as value
+				continue
+			}
+			diags = append(diags, c.evaluateLocalVariable(local)...)
+			copy(locals[i:], locals[i+1:])
+		}
+	}
+
+	// Evaluate locals containing another local variable(s)
+	for _, local := range locals {
+		diags = append(diags, c.evaluateLocalVariable(local)...)
+	}
+
+	return diags
```

Comment:
Is there a way to prevent a panic in the case someone does something like this ( or with more variables in the middle) :

```hcl
locals {
  first  = local.second
  second = local.first
}
```

I think we might need to make a tree out of those. Or may be something simpler ? Like a sort.Slice, sorting by dependencies and variable name ?

Reply:
By saying panic you mean any error or panic itself? 🤔 this returns an error like it couldn't find the variables 
```
This object does not have an attribute named "second".

Error: Unsupported attribute

  on variables.pkr.hcl line 23, in locals:
  23:   second = local.first

This object does not have an attribute named "first".

Error: Unsupported attribute

  on variables.pkr.hcl line 23, in locals:
  23:   second = local.first

This object does not have an attribute named "first".

Error: Unsupported attribute

  on variables.pkr.hcl line 23, in locals:
  23:   second = local.first

This object does not have an attribute named "first".

Error: Unsupported attribute

  on variables.pkr.hcl line 23, in locals:
  23:   second = local.first

This object does not have an attribute named "first".
```

I thought this was supposed to happen 🤔

Reply:
Now I'm questioning myself if this would work with something like: 
```
```hcl
locals {
  first  = "first"
  second = local.first
  third = local.second_loca.first
}